### PR TITLE
Bugfix #350 and some fixes in RStarTreeSplit and XTreeSplit

### PR DESCRIPTION
**Pull request by lozhnikov**
XTreeSplit::InsertNodeIntoTree uses std::vector::push_back which increases the size of children vector.

bestAreaIndexOnBestAxis and bestOverlapIndexOnBestAxis should not be redefined in SplitNonLeafNode.

Otherwise the global variables will be equal 0.

### CODE REVIEW

**Comment by rcurtin** [comment]
Thanks for the contribution.  One of my concerns is that you are using the `RTreeDescentHeuristic` class instead of the `XTreeDescentHeuristic` class, which wasn't ever implemented (it was an unfinished GSoC project from 2014).  This is only doable if the X-tree (as specified in the paper) uses the same descent heuristic as the original R tree.  Do you know if that's the case?

Thanks for the RStarTreeSplit fixes too, those must have been oversights when I was refactoring.

**Comment by lozhnikov** [comment]
I looked through the X-tree paper. The authors don't explain the choosing subtree algorithm. But according to the R*-tree paper, alternative methods didn't outperform Guttman's original algorithm (for choosing the best non-leaf node) and for the leaf nodes, minimizing the overlap is better.

However,  the main idea of X-tree is to reduce overlap in the split algorithm. Why do not use the same idea? For example, 
1. Try the R* algorithm.
2. If the overlap cost > MAX_OVERLAP, try to choose  the node with the minimum enlargement and overlap cost < MAX_OVERLAP.

But it is my assumption, I haven't any benchmarks yet.

**Comment by rcurtin** [comment]
Ok, great.  Give me a few days to read the paper and verify that; mostly I just wanted to make sure that this PR considered that.  The PR looks good otherwise, and I'm happy to have another tree to list in the trees we implement (I guess if you want you can add the X tree as an option to programs like `src/mlpack/methods/neighbor_search/allknn_main.cpp` and others).  Also, would you like to also update `src/mlpack/core.hpp` and `COPYRIGHT.txt` and add your name to the list of contributors?

**Comment by lozhnikov** [comment]
I added x-trees to allkfn, allknn, range_search and allkrann. X-trees show poor performance (maybe something is wrong). Thanks for your offer, but I think my contribution is not substantial, maybe later.

**Comment by rcurtin** [comment]
Thanks for the `lowest()` fix, I'm not sure how I misunderstood the documentation for that.

I tried running `mlpack_allknn` with your branch to test the speed but it segfaulted.  So I recompiled with debugging symbols, and ran it through valgrind, and saw this:

```
==24731== Invalid read of size 8
==24731==    at 0x989CF8: bool mlpack::tree::XTreeSplit::SplitNonLeafNode<mlpack::tree::RectangleTree<mlpack::metric::LMetric<2, true>, mlpack::neighbor::NeighborSearchStat<mlpack::neighbor::NearestNeighborSort>, arma::Mat<double>, mlpack::tree::XTreeSplit, mlpack::tree::RStarTreeDescentHeuristic> >(mlpack::tree::RectangleTree<mlpack::metric::LMetric<2, true>, mlpack::neighbor::NeighborSearchStat<mlpack::neighbor::NearestNeighborSort>, arma::Mat<double>, mlpack::tree::XTreeSplit, mlpack::tree::RStarTreeDescentHeuristic>*, std::vector<bool, std::allocator<bool> >&) (x_tree_split_impl.hpp:811)
(plus lots more output)
```

I haven't had a chance to look into it yet, and probably won't for a little while, unfortunately.

A last thought---adding the support for X trees to NSModel and RSModel and RAModel is great (thanks!) but we should add the X tree after the ball tree in order to preserve reverse compatibility.  Otherwise old models that were built with a ball tree will be loaded as an X tree (and the load will fail) because X_TREE will be equal to what BALL_TREE was in older versions.

I'm excited about merging this functionality in, so thanks for taking the time to look into this!  If you don't want to add your name that's fine, but personally I think your contribution is pretty substantial. Usually the level for inclusion in the credits is just "more than one line of code". :)

**Comment by lozhnikov** [comment]
I found some faults in x-trees split when numChildren > maxNumChildren or children.size() != maxNumChildren+1. These errors can happen due to splitting a supernode or removing a node in CondenseTree, it leads to segfault.

Do you know why ball trees aren't added to allkran? If I am not mistaken it is forgotten.

**Comment by rcurtin** [comment]
I read through the X tree paper again.  I agree that the DescentType is not specified... the paper just says in Figure 7:

```
follow = choose_subtree(obj);
```

And there is no explanation of what `choose_subtree()` is, so I think it's safe to assume that it's the typical R tree heuristic.  (Maybe it would be possible to use the R* heuristic too, but, I think sticking with the default is good for the X tree typedef.)

Ball trees aren't added to allkrann because of a tricky bug:
https://github.com/mlpack/mlpack/issues/338

**Comment by lozhnikov** [comment]
I've found another bug in r/r*/x trees. But I am not sure in the best solution.

Suppose you use the RectangleTree(const RectangleTree& other,const bool deepCopy) constructor.
(it is used in all Split methods)
It results in a new dataset. 
`    dataset(new MatType(*other.dataset)),
    ownsDataset(true),
`
Therefore we should not delete this node until we delete all its children which are created using the RectangleTree(Rectangle tree *parentNode) constructor.
But it may happen (and it happens) in split methods and in RectangleTree::CondenseTree.

Possible solutions:
1. Fix constructor
`    dataset(deepCopy ? new MatType(*other.dataset) : &other->Dataset()),
    ownsDataset(deepCopy),
`
But I am not sure of deepCopy=true (Is it used anywhere? In this case a deep copy should be a root)
2. Why do you use
`const MatType* dataset;` ?
Maybe it is better to use something like this
`MatType *dataset;
 MatType* DatasetPtr() const { return dataset; }
 MatType*& DatasetPtr() { return dataset; }
`
But it results in a recursive change of all pointers in all descendant nodes.

**Comment by lozhnikov** [comment]
I think 121bbafe14eba62873f20271ea59612b7b5bcae9 should fix the problem explained above.

**Comment by rcurtin** [comment]
Sorry for the slow response, I'm a bit overwhelmed at the moment.  I will try to get back to this today or tomorrow, but it may be Monday until I can (the same applies to your excellent CosineTree fix too).

**Comment by rcurtin** [comment]
I'm really sorry this took so long.  The end of the GSoC application period was pretty overwhelming and then I went on vacation for a week, so only now am I able to take a look at this in-depth.

Here is some background based on what I remember of this during GSoC 2014.  The rectangle tree and its variants are made for dynamic point sets where insertions and deletions are common, but this is a little bit at odds with the other trees in mlpack, which tend to be "build it once and don't insert or delete points from the dataset".  In order to support this design decision, the original idea was to have each node hold its own arma::mat (this is called `localDataset`) and then points could easily be inserted or deleted by just calling `insert_cols()` or `shed_cols()` on the local datasets.  But I am not sure this ever really happened in earnest, so I think maybe the `Insert()` and `Delete()` functions do not work correctly.  This is a thing I would eventually like to look into, but have not yet had time to.  I don't think the local dataset is even used; instead I think each node holds a `std::vector<size_t>` that holds the indices of the points that are in that node.  If you are interested in solving that problem and making the Insert()/Delete() support meaningful, I can discuss it more if you like.

The other issue was the necessity for a shallow copy constructor, because sometimes the insertion procedure necessitates creating a new root node.  But if we are constructing the tree like

```
RTreeType* tree = new RTreeType(dataset);
```

we can't have `tree` point to an intermediate level of the tree; it must point to the root.  So we need to make a shallow copy of the current `tree` node, then make that shallow copy a child of the current `tree` node and update all the data in the `tree` node.

I think you are right that you uncovered a problem there, and I think that your solution in 121bbaf is the right solution.  The situation where `deepCopy = true` is the default copy constructor behavior, and would be used if you wanted to copy the tree entirely:

```
RTreeType myTree(dataset);
RTreeType myTreeCopy(myTree /* defaults to a deep copy */);
```
For the second issue, the `const MatType* dataset` only makes sense if the tree should not be modifying the dataset.  We should just remove the non-const function `Dataset()`, to be honest.

File: src/mlpack/methods/neighbor_search/ns_model_impl.hpp [hunk 1]
```diff
@@ -464,6 +498,8 @@ std::string NSModel<SortPolicy>::TreeName() const
       return "R tree";
     case R_STAR_TREE:
       return "R* tree";
+    case X_TREE:
+      return "X tree";
```
Comment (rcurtin) [inline src/mlpack/methods/neighbor_search/ns_model_impl.hpp]: If you don't mind, when I merge this I'll probably just change all of the x-tree cases to come after the ball tree, like in the enum.  It's a really trivial change but it preserves the ordering and makes it easier to read, in my opinion.

File: src/mlpack/core/tree/rectangle_tree/rectangle_tree.hpp [hunk 2]
```diff
@@ -76,6 +76,9 @@ class RectangleTree
   } SplitHistoryStruct;
 
  private:
+  //! The max number of child nodes a non-leaf normal node can have. 
+  //! (used in x-trees)
+  size_t normalNodeMaxNumChildren;
```
Comment (rcurtin) [inline src/mlpack/core/tree/rectangle_tree/rectangle_tree.hpp]: I don't like putting X tree logic in the generic RectangleTree class; I think maybe the right thing here is some refactoring.  I wonder if maybe `CondenseTree()` should call some function in `SplitType`.  But probably the refactoring that has to happen to make this work is pretty significant, so if you can't think of something now we can open a ticket and save it for some other time.  I read through `CondenseTree()` but it was only a quick read and I didn't think of any easy solutions for refactoring this x-tree logic out.  Any ideas?

**Comment by rcurtin** [comment]
The code looks good to me; thanks for updating the tests and everything.  Let me know what you think about the comment I made about the X-tree specific code and we can merge this now and leave that for later, or keep working on that issue now.

**Comment by lozhnikov** [comment]
I see two posiibilities:
1. I can fix `SplitNonLeafNode()` in such a way as to work up the case when a node has only one child and has'not got the parent (i.e. this is the root). So `CondenseTree()` should invoke this method in the case discussed.
2. The second way is to create a new method in `SplitType` e.g. `CondenseToRoot()` and invoke this method in `CondenseTree()`.

I think the second one is better. The `normalNodeMaxNumChildren` was introduced in order to reduce memory usage since the leaf nodes do not contain children and the parent of a leaf node can be a supernode. So if I split a leaf node, the resulting nodes should not allocate the space like the parent of the leaf node.

What do you think about it?

**Comment by rcurtin** [comment]
I agree that the second approach is better here.

Unless I have misunderstood, `normalModeMaxNumChildren` is something specific to the X tree, so I think we should try to remove the member from `RectangleTree` and put it in `XTreeSplit` if possible.  I think that this is reasonable, but will require some refactoring (so we can save it for later if you want):

We'd need to make `RectangleTree` use instantiated split objects and (optionally) descent type objects.  So instead of `SplitType::SplitNode()`, `RectangleTree` should hold a `SplitType split` object (or something like this), and then call `split.SplitNode()` instead.  That could allow the SplitType to hold state (like for instance the value of `normalModeMaxNumChildren`), and then the splitting function could make use of that state, and the other R-tree variants wouldn't need to hold something like `normalModeMaxNumChildren`.

Let me know what you think; hopefully I have understood everything correctly.  Like I said I'm happy to leave `normalModeMaxNumChildren` in the code for now and then open a ticket noting that it should be refactored at some point later in the future.

**Comment by lozhnikov** [comment]
That's right. I think I'll fix `SplitType` in the next week.

**Comment by rcurtin** [comment]
Sounds good, thanks!

**Comment by lozhnikov** [comment]
I replaced the static `SplitType` class by an instantiated object.
The `normalNodeMaxNumChildren` was moved to the `XTreeSplit` class.

If I am not mistaken the `splitHistory` also should be moved to the `XTreeSplit` class.

**Comment by rcurtin** [comment]
I agree, `splitHistory` should definitely be part of `XTreeSplit`.  The refactoring looks great to me; my only comment is, I think if the `SplitType` classes hold a pointer to the node they are associated with, this may increase the size of the `RectangleTree` classes.  So I think maybe it would be better to have each of the `SplitType` methods take a pointer to the node, instead of taking the pointer to the node at construction time.  Maybe I have overlooked some reason why that isn't possible---let me know if I've made an error in my thinking.

I'd like to go ahead and merge this for now, so I can release mlpack 2.0.2 with these nice fixes.  Is that okay?  We can finish with the other issues in other PRs as needed.

**Comment by lozhnikov** [comment]
Yes, it is possible. I think I'll fix the pointers in the `SplitType` class and I'll move `splitHistory` tomorrow. So if you will merge before that, I'll create a new PR.

**Comment by rcurtin** [comment]
I can wait for that, a day (or two) is no issue.  I still have to find time to prepare the release; maybe later this week I hope. :)

**Comment by lozhnikov** [comment]
I fixed the issues discussed above. But I am not sure of `SplitType::SplitType(const TreeType *node)` and `SplitType::SplitType(const TreeType *node,const TreeType *parent)`. Maybe these constructors should be merged into one.

**Comment by rcurtin** [comment]
Yeah, we can just combine those two constructors, since `node->Parent()` can always be used to get the node's parent if needed.

**Comment by lozhnikov** [comment]
Well, I don't see any issues in this code. So, I think it's time to merge it.

**Pull request merged**